audit.md

```diff
@@ -0,0 +1,212 @@
+    - Will wait until `aragon-apps` is upgraded to `aragonOS 3.0.0` before running Oyente and unit tests as some contracts in `aragonOS` have moved in `3.0.0` (causing compile issues)
+- Slight, non-logical modifications were made to some contracts for running [Solgraph](#solgraph) (see [cd431a899af22b6f67787a8fc6cf7e1c60f8a6a7](https://github.com/sohkai/aragon-apps/commit/cd431a899af22b6f67787a8fc6cf7e1c60f8a6a7)); **these changes should not be adopted**
+- Slight, non-logical modifications were made to some contracts for running [Oyente](#oyente) (see [475b72f81c1a6fd136391ff1bf3a93228fb57d91](https://github.com/sohkai/aragonOS/commit/475b72f81c1a6fd136391ff1bf3a93228fb57d91)); **these changes should not be adopted**
+
+
+## General comments
+
+- No linting errors
+- All unit tests passing (`aragonOS` only; see note above)
+- Uses industry-standard library contracts for common operations, including OpenZeppelin and MiniMe
+
+### Out of scope
+
+I have, for the most part, ignored:
+
+- Any contracts in `lib` folders
+- Any contracts in `factory` folders (they're mostly for demonstration right now)
+- An exhaustive look through the unit tests to see if they miss any obvious functionality (100% code and branch coverage is a good start but is usually not the whole story)
+
+
+## Issues
+
+- As a scale, I'm using:
+    - High-severity: anything that can result in loss of funds for users, or otherwise achieve highly damaging effects on user DAOs
+    - Mid-severity: exploitable issues, but either preventable in some way, technically difficult (or impossible) to solve for every one in the ecosytem, or very low probability of actually being exploited
+    - Low-severity: nice-to-haves, very low incentive attacks
+
+### Mid-severity
+
+- `AppProxy` should fail on construction if `_kernel` is not a contract
+    - `aragonOS/contracts/apps/AppProxy.sol`:
+        - Kernel-less apps are impossible: the fallback function will always fail on `kernel.getAppCode(appId)` if `kernel == 0` or if the kernel address is not a contract.
+    - `aragonOS/contracts/apps/AragonApp.sol`:
+        - `canPerform()` would not require the `address(kernel) == 0` check anymore as the proxy would fail to be created before that point
+    - `aragonOS/contracts/apm/ApmRegistry.sol`:
+        - `initialize()`'s `require(kernel != address(0));` check can be removed
+- `AppProxy` may use garbage data as its `target`:
+    - `aragonOS/contracts/apps/AppProxy.sol`:
+        - If the `AppProxy`'s `kernel` is set to a non-kernel contract address, `kernel.getAppCode(appId)` will return garbage data [due to how the fallback works](https://github.com/ethereum/solidity/issues/2630). This means that in `delegatedFwd(target, msg.data)`, `target` could potentially be delegating to a malicious contract. As far as I know, exploiting this would be extremely difficult, but it should be clear in the surrounding documentation that the `kernel` **MUST** be carefully set.
```

Comment:
This is a fantastic point indeed. I would separate the scenario in which this happens by mistake (and the garbage return is a problem), which can be easily solved by checking for example a role in the kernel (though at the expense of an extra ~600 gas for the call).

In the case this is done maliciously, I don't think there is a solution that doesn't involve an Aragon 'approved kernel bytecode' which would be super expensive. 

I would solve this with red big fat warnings in the client if we detect something like this. Also another point, is that the client won't recognize a contract as a DAO's app unless the kernel address is the DAO's kernel address.

Reply:
> Red big fat warnings in the client

Yes, in general, any app that gets loaded (from the created permission set) but whose kernel address does not match the DAO's kernel address should show huge warnings.